src/libtiled/mapreader.cpp

```diff
@@ -1453,6 +1459,42 @@ void MapReaderPrivate::readProperty(Properties *properties)
     properties->insert(propertyName, variant);
 }
 
+Components MapReaderPrivate::readComponents()
+{
+  Q_ASSERT(xml.isStartElement() && xml.name() == QLatin1String("components"));
+
+  Components commponents;
+
+  while (xml.readNextStartElement()) {
+    if (xml.name() == QLatin1String("component"))
+      readComponent(&commponents);
+    else
+      readUnknownElement();
+  }
+
+  return commponents;
+}
+
+void MapReaderPrivate::readComponent(Components *components)
+{
+  Q_ASSERT(xml.isStartElement() && xml.name() == QLatin1String("component"));
+
+  const QXmlStreamAttributes atts = xml.attributes();
+  QString componentName = atts.value(QLatin1String("name")).toString();
+
+  while (xml.readNext() != QXmlStreamReader::Invalid) {
```

Comment:
Why not `while (xml.readNextStartElement()) {` here as well?

Reply:
I mirrored the behaviour of `readProperty()`.

Reply:
Ah, but `readProperty` can't use `readNextStartElement` because it needs to also handle the `xml.isCharacters` case.